node/src/identity/hsmsigners.ts

```diff
@@ -0,0 +1,186 @@
+import { Signer } from './signer';
+
+type DER = string | number[];
+
+export interface HSMSignerOptions {
+    /**
+     * the path to the PKCS11 library
+     */
+    library: string;
+
+    /**
+     * The label associated with the token for the slot
+     */
+    label: string;
+
+    /**
+     * The pin for the slot identified by the label
+     */
+    pin: string;
+
+    /**
+     * Identifier. The CKA_ID assigned to the HSM object
+     */
+    identifier: string | Buffer;
+
+    /**
+     * Optional user type for the HSM. If not specified it defaults to CKU_USER
+     */
+    userType?: number;
+}
+
+export type HSMClose = () => void;
+export type HSMSigner = {signer: Signer, close: HSMClose};
+
+let hsmSignerFactory: HSMSignerFactory | undefined;
+
+/**
+ * Create a new HSM signing implementation based on provided HSM options.
+ *
+ * This returns an object with 2 properties
+ * - signer which is the signer function
+ * - close which is a close function to close the signer when it's not required anymore
+ *
+ * @param hsmSignerOptions - The HSM signer options
+ * @returns an HSM Signer implementation
+ */
+export const newHSMSigner = (hsmSignerOptions: HSMSignerOptions): HSMSigner => {
+    if (!hsmSignerFactory) {
+
+        if (!hsmSignerOptions.library || hsmSignerOptions.library.trim() === '') {
+            throw new Error('library property must be provided');
+        }
+
+        hsmSignerFactory = new HSMSignerFactory(hsmSignerOptions.library);
+    }
+
+    if (!hsmSignerOptions.label || hsmSignerOptions.label.trim() === '') {
+        throw new Error('label property must be provided');
+    }
+
+    if (!hsmSignerOptions.pin || hsmSignerOptions.pin.trim() === '') {
+        throw new Error('pin property must be provided');
+    }
+
+    if (!hsmSignerOptions.identifier || hsmSignerOptions.identifier.toString().trim() === '') {
+        throw new Error('identifier property must be provided');
+    }
+
+    const supportedKeySize = 256;
+
+    return hsmSignerFactory.newSigner(hsmSignerOptions.label, hsmSignerOptions.pin, hsmSignerOptions.userType, supportedKeySize, hsmSignerOptions.identifier);
+}
+
+class HSMSignerFactory {
+
+    #pkcs11: pkcs11js.PKCS11;
+
+    constructor(library: string) {
+        this.#pkcs11 = new pkcs11js.PKCS11()
+        this.#pkcs11.load(library);
+        this.#pkcs11.C_Initialize();
+    }
+
+    public newSigner(label: string, pin: string, userType: number = pkcs11js.CKU_USER, keySize: number, identifier: string | Buffer): {signer: Signer, close: () => void} {
+        const pkcs11 = this.#pkcs11;
+        const slot = this.findSlotForLabel(label);
+        const session = pkcs11.C_OpenSession(slot, pkcs11js.CKF_SERIAL_SESSION);
+        pkcs11.C_Login(session, userType, pin);
+        const privateKeyHandle = this.findObjectInHSM(session, pkcs11js.CKO_PRIVATE_KEY, identifier);
+
+        const curveName = `secp${keySize}r1`;
```

Comment:
This is only used for error messages. Is it worth losing this and just using the actual curve name we ask for (`p${keySize}`) instead?